csharp/src/Apache.Arrow/RecordBatch.Builder.cs

```diff
@@ -0,0 +1,149 @@
+            public Int8Array Int8(Action<Int8Array.Builder> action) => Build<Int8Array, Int8Array.Builder>(action);
+            public Int16Array Int16(Action<Int16Array.Builder> action) => Build<Int16Array, Int16Array.Builder>(action);
+            public Int32Array Int32(Action<Int32Array.Builder> action) => Build<Int32Array, Int32Array.Builder>(action);
+            public Int64Array Int64(Action<Int64Array.Builder> action) => Build<Int64Array, Int64Array.Builder>(action);
+            public UInt8Array UInt8(Action<UInt8Array.Builder> action) => Build<UInt8Array, UInt8Array.Builder>(action);
+            public UInt16Array UInt16(Action<UInt16Array.Builder> action) => Build<UInt16Array, UInt16Array.Builder>(action);
+            public UInt32Array UInt32(Action<UInt32Array.Builder> action) => Build<UInt32Array, UInt32Array.Builder>(action);
+            public UInt64Array UInt64(Action<UInt64Array.Builder> action) => Build<UInt64Array, UInt64Array.Builder>(action);
+            public FloatArray Float(Action<FloatArray.Builder> action) => Build<FloatArray, FloatArray.Builder>(action);
+            public DoubleArray Double(Action<DoubleArray.Builder> action) => Build<DoubleArray, DoubleArray.Builder>(action);
+            public Date32Array Date32(Action<Date32Array.Builder> action) => Build<Date32Array, Date32Array.Builder>(action);
+            public Date64Array Date64(Action<Date64Array.Builder> action) => Build<Date64Array, Date64Array.Builder>(action);
+            public BinaryArray Binary(Action<BinaryArray.Builder> action) => Build<BinaryArray, BinaryArray.Builder>(action);
+            public StringArray String(Action<StringArray.Builder> action) => Build<StringArray, StringArray.Builder>(action);
+
+            private TArray Build<TArray, TArrayBuilder>(Action<TArrayBuilder> action)
+                where TArray: IArrowArray
+                where TArrayBuilder: IArrowArrayBuilder<TArray>, new()
+            {
+                if (action == null)
+                {
+                    return default;
+                }
+
+                var builder = new TArrayBuilder();
+                action(builder);
+
+                return builder.Build(_memoryPool);
+            }
+        }
+
+        public class Builder
+        {
+            private readonly MemoryPool _memoryPool;
+            private readonly Schema.Builder _schema;
+            private readonly LinkedList<IArrowArray> _arrays;
+
+            public Builder(MemoryPool memoryPool = default)
+            {
+                _memoryPool = memoryPool ?? MemoryPool.Default.Value;
+                _schema = new Schema.Builder();
+                _arrays = new LinkedList<IArrowArray>();
+            }
+
+            public RecordBatch Build()
+            {
+                var schema = _schema.Build();
+                var length = _arrays.Max(x => x.Length);
+                var batch = new RecordBatch(schema, _arrays, length);
```

Comment:
If you use the new ctor I'm adding here:

[https://github.com/apache/arrow/pull/3736/files#diff-96784cd3353fe0069df2887812e5b784R56](https://github.com/apache/arrow/pull/3736/files#diff-96784cd3353fe0069df2887812e5b784R56)

Then you can avoid allocating an extra List and copying the instances over.